lib/xnvme_be_spdk_async.c

```diff
@@ -146,13 +146,97 @@ xnvme_be_spdk_async_cmd_io(struct xnvme_cmd_ctx *ctx, void *dbuf, size_t dbuf_nb
 
 	return 0;
 }
+
+static void
+reset_iov_offset(void *cb_arg, uint32_t offset)
+{
+	struct xnvme_cmd_ctx *ctx = cb_arg;
+	struct xnvme_be_spdk_state *state = (void *)ctx->dev->be.state;
+	struct iovec *iov;
+
+	state->iov_offset = offset;
+
+	for (state->iov_offset = 0; state->iov_pos < state->iov_cnt; state->iov_pos++) {
+		iov = &state->iov[state->iov_pos];
+		if (state->iov_offset < iov->iov_len) {
+			break;
+		}
+
+		state->iov_offset -= iov->iov_len;
+	}
+}
+
+static int
+next_iov_entry(void *cb_arg, void **address, uint32_t *length)
+{
+	struct xnvme_cmd_ctx *ctx = cb_arg;
+	struct xnvme_be_spdk_state *state = (void *)ctx->dev->be.state;
+	struct iovec *iov;
+
+	iov = &state->iov[state->iov_pos];
+
+	*address = iov->iov_base;
+	*length = iov->iov_len;
+
+	if (state->iov_offset) {
+		*address += state->iov_offset;
+		*length -= state->iov_offset;
+	}
+
+	state->iov_offset += *length;
+	if (state->iov_offset == iov->iov_len) {
+		state->iov_pos++;
+		state->iov_offset = 0;
+	}
+
+	return 0;
+}
+
+int
+xnvme_be_spdk_async_cmd_iov(struct xnvme_cmd_ctx *ctx, struct iovec *dvec, size_t dvec_cnt,
+			    size_t dvec_nbytes, struct iovec *mvec, size_t mvec_cnt,
+			    size_t XNVME_UNUSED(mvec_nbytes))
+{
+	struct xnvme_queue_spdk *queue = (void *)ctx->async.queue;
+	struct xnvme_be_spdk_state *state = (void *)queue->base.dev->be.state;
+	void *mbuf = NULL;
+
+	int err;
+
+	if (mvec_cnt > 1) {
+		XNVME_DEBUG("FAILED: mvec_cnt must be 1 at most");
+		return -EINVAL;
+	}
+
+	if (mvec_cnt) {
+		mbuf = mvec[0].iov_base;
+	}
+
+	state->iov = dvec;
+	state->iov_cnt = dvec_cnt;
+	state->iov_pos = 0;
+	state->iov_offset = 0;
+
+	queue->base.outstanding += 1;
+
+	err = spdk_nvme_ctrlr_cmd_iov_raw_with_md(
```

Comment:
We need to double-check the lifetime of the attributes accessed on ``state``. E.g. if it is used **after** the return of ``spdk_nvme_ctrlr_cmd_iov_raw_with_md()``, then we will have races on the structure.